informal_update/serializers.py

```diff
@@ -0,0 +1,117 @@
+from django.utils.translation import ugettext
+
+from rest_framework import serializers
+
+from enumfields.drf.serializers import EnumSupportSerializerMixin
+
+from api.serializers import (
+    UserNameSerializer,
+    DisasterTypeSerializer,
+    CountrySerializer,
+    MiniDistrictSerializer,
+)
+from informal_update.models import (
+    InformalReferences,
+    InformalUpdate,
+    InformalCountryDistrict,
+    InformalGraphicMap,
+    InformalAction,
+    InformalActionsTaken
+)
+
+from main.writable_nested_serializers import (
+    NestedCreateMixin,
+    NestedUpdateMixin
+)
+
+
+class InformalGraphicMapSerializer(serializers.ModelSerializer):
+    created_by_details = UserNameSerializer(source='created_by', read_only=True)
+
+    class Meta:
+        model = InformalGraphicMap
+        fields = '__all__'
+        read_only_fields = ('created_by',)
+
+    def create(self, validated_data):
+        validated_data['created_by'] = self.context['request'].user
+        return super().create(validated_data)
+
+
+class InformalActionSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = InformalAction
+        fields = ('name', 'id', 'organizations', 'category', 'tooltip_text')
+
+
+class InformalActionsTakenSerializer(
+    NestedUpdateMixin,
+    NestedCreateMixin,
+    serializers.ModelSerializer
+):
+    action_details = InformalActionSerializer(source='actions', many=True, required=False, read_only=True)
+    organization_display = serializers.CharField(source='get_organization_display', read_only=True)
+
+    class Meta:
+        model = InformalActionsTaken
+        fields = ('organization', 'organization_display', 'actions', 'action_details', 'summary', 'id',)
+        read_only_fields = ('informal_update',)
+
+
+class InformalReferencesSerializer(
+    serializers.ModelSerializer
+):
+    document_details = InformalGraphicMapSerializer(source='document', read_only=True)
+
+    class Meta:
+        model = InformalReferences
+        fields = '__all__'
+
+
+class InformalCountryDistrictSerializer(serializers.ModelSerializer):
+    country_details = CountrySerializer(source='country', read_only=True)
+    district_details = MiniDistrictSerializer(source='district', read_only=True)
+
+    class Meta:
+        model = InformalCountryDistrict
+        fields = ('id', 'country', 'district', 'country_details', 'district_details')
+        read_only_fields = ('informal_update',)
+
+    def validate(sel, data):
+        if len(data) > 10:
+            raise serializers.ValidationError("Number of countries selected should be less than 10")
+        district = data['district']
+        if district and district.country != data['country']:
+            raise serializers.ValidationError({
+                'district': ugettext('Different districts found for given country')
+            })
+        return data
```

Comment:
Lets add field level validations for these

Reply:
If we do field level validation like validate_district(self, data), then data is treated as object <class 'api.models.District'>
, so getting district = data['district'] leads to error TypeError: 'District' object is not subscriptable. Getting data['country'] also leads to error. So lets keep it as it is.

Reply:
This is okay since multiple field values (country and district) are used.